runtime/src/bank/builtins/core_bpf_migration/source_bpf_upgradeable.rs

```diff
@@ -0,0 +1,365 @@
+use {
+    super::error::CoreBpfMigrationError,
+    crate::bank::Bank,
+    solana_sdk::{
+        account::{AccountSharedData, ReadableAccount},
+        bpf_loader_upgradeable::{
+            get_program_data_address, UpgradeableLoaderState, ID as BPF_LOADER_UPGRADEABLE_ID,
+        },
+        pubkey::Pubkey,
+    },
+};
+
+/// Used to validate a source BPF Upgradeable program's account and data
+/// account before migrating a built-in program to Core BPF.
+#[derive(Debug)]
+pub(crate) struct SourceProgramBpfUpgradeable {
+    pub program_address: Pubkey,
+    pub program_account: AccountSharedData,
+    pub program_data_address: Pubkey,
+    pub program_data_account: AccountSharedData,
+    pub total_data_size: usize,
```

Comment:
Are you just trying to save space in memory here? It seems oddly nonspecific to add the two account data-sizes together, and leaves checks vulnerable to cases where the two accounts increase/decrease by the same amount, respectively. How about we just store the two lengths in separate fields?

Reply:
I keep this value to make the calculation for the account data size delta more straightforward: https://github.com/anza-xyz/agave/pull/79/files#diff-9e66434292cab4fc58e936cb5da74f3ab209fb660716a6fdba7546d2b922b52fR104-R108

I can store the two lengths separately, but then it kind of defeats the purpose of keeping both values separate when you could just call `.data.len()` on both accounts in the struct.

Reply:
Maybe I'm missing something on how that delta calculation is supposed to go? I adopted it from `Bank::replace_program_account`.

Reply:
Just for some additional context:
https://github.com/anza-xyz/agave/pull/408/files#diff-9e66434292cab4fc58e936cb5da74f3ab209fb660716a6fdba7546d2b922b52fR102-R116

Reply:
Oh, this is only used for `calculate_and_update_accounts_data_size_delta_off_chain()`? I was envisioning sanity checks that the account data remained as expected. Sorry for not looking at the actual usage example.

Now that I see it, I don't think there's much point to carrying around `SourceUpgradeableBpf::total_data_size` or `TargetBuiltin::total_data_size`. I think remove those fields and just call `AccountSharedData::data().len()` on each account at the call site.

Reply:
Sounds good I let that field go!